### PR TITLE
Refactor recommendation to an interface

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

### CODE REVIEW

File: pkg/cluster/k8s.go [hunk 1]
```diff
@@ -35,199 +28,31 @@ func NewK8sProvisioner() clusterTypes.Provisioner {
...
-		return nil, nil, errors.New("failed to create namespace " + recommend.NS)
-	}
-	err = k.ABTest.Apply(recommend)
-	if err != nil {
-		return nodes, clientNodes, err
-	}
-	nodes, err = k.ABTest.GetNodes(recommend)
-	if err != nil {
-		return nodes, clientNodes, err
+	// TODO: add iochaos to the new interface
```
Comment (mahjonp) [inline pkg/cluster/k8s.go:31]: If you like some configs could be pass down to sub cluster easily, maybe ClusterSpecs can help you. eg. pass it to Apply method.

File: pkg/test-infra/tiflash/operation.go [hunk 2]
```diff
@@ -25,60 +25,65 @@ import (
...
+// Namespace ...
+func (o *Ops) Namespace() string {
+	return o.ns
 }
 
 // Apply TiFlash cluster
-func (o *Ops) Apply(tc *Recommendation) error {
-	if err := o.TidbOps.ApplyTiDBCluster(tc.TiDBCluster); err != nil {
+func (o *Ops) Apply() error {
+	if err := o.applyObject(o.tf.ConfigMap); err != nil {
```
Comment (mahjonp) [inline pkg/test-infra/tiflash/operation.go:54]:
```suggestion
	if err := util.ApplyObject(o.cli, o.tf.ConfigMap); err != nil {
```

File: pkg/test-infra/tiflash/operation.go [hunk 3]
```diff
@@ -25,60 +25,65 @@ import (
...
 		return err
 	}
-	if err := o.ApplyTiFlash(tc.TiFlash); err != nil {
+	if err := o.applyObject(o.tf.Service); err != nil {
```
Comment (mahjonp) [inline pkg/test-infra/tiflash/operation.go:57]:
```suggestion
	if err := util.ApplyObject(o.cli, o.tf.ConfigMap); err != nil {
```

File: pkg/test-infra/tiflash/operation.go [hunk 4]
```diff
@@ -25,60 +25,65 @@ import (
...
 		return err
 	}
-	return nil
-}
-
-// Delete TiFlash cluster
-func (o *Ops) Delete(tc *Recommendation) error {
-	if err := o.TidbOps.Delete(tc.TiDBCluster); err != nil {
+	if err := o.applyObject(o.tf.StatefulSet); err != nil {
```
Comment (mahjonp) [inline pkg/test-infra/tiflash/operation.go:60]:
```suggestion
	if err := util.ApplyObject(o.cli, o.tf.ConfigMap); err != nil {
```

File: pkg/test-infra/tiflash/operation.go [hunk 5]
```diff
@@ -25,60 +25,65 @@ import (
...
 		return err
 	}
 
-	if err := o.DeleteTiFlash(tc.TiFlash); err != nil {
+	if err := o.waitTiFlashReady(o.tf.StatefulSet, 5*time.Minute); err != nil {
```
Comment (mahjonp) [inline pkg/test-infra/tiflash/operation.go:64]:
```suggestion
	if err := o.waitTiFlashReady(o.tf.StatefulSet, fixture.Context.WaitClusterReadyDuration); err != nil {
```

**Comment by yeya24** [comment]
I have updated my code. Please give a review when you are free @mahjonp  @cwen0  @you06

File: pkg/core/nemesis.go [hunk 6]
```diff
@@ -25,9 +25,9 @@ const (
 	// TimeChaos means
 	TimeChaos ChaosKind = "time-chaos"
 	// PDScheduler adds scheduler
-	PDScheduler ChaosKind = "pd-scheduler"
+	PDScheduler ChaosKind = "PD-Scheduler"
 	// PDLeaderShuffler will randomly shuffle pds.
-	PDLeaderShuffler ChaosKind = "pd-leader-shuffler"
+	PDLeaderShuffler ChaosKind = "PD-Leader-Shuffler"
```
Comment (mahjonp) [inline pkg/core/nemesis.go:30]: use lower case here?

File: pkg/pocket/pkg/mysql/mysql.go [hunk 7]
```diff
@@ -81,7 +81,7 @@ func (conn *DBConnect) IfTxn() bool {
 	return conn.txn != nil
 }
 
-// GetTiDBTS get the txn begin timestamp from TiDB
+// GetTiDBTS get the txn begin timestamp from TiD
```
Comment (mahjonp) [inline pkg/pocket/pkg/mysql/mysql.go:84]:
```suggestion
// GetTiDBTS get the txn begin timestamp from TiDB
```

File: pkg/test-infra/tiflash/recommend.go [hunk 8]
```diff
@@ -59,27 +49,32 @@ func RecommendedTiFlashCluster(ns, name, version string) *Recommendation {
...
-		TiDBCluster: tc,
-		TiFlash: &TiFlash{
-			StatefulSet: tiFlashStatefulSet(tiFlashName, lbls, model),
-			// we use name instead of tiFlashName here
-			// because we want to use it to do template rendering.
-			ConfigMap: tiFlashConfigMap(name, model),
-			Service:   tiFlashService(ns, tiFlashName, lbls),
-		},
+
+	// (TODO: yeya24) Unfortunately we have to parse the nemesis
```
Comment (mahjonp) [inline pkg/test-infra/tiflash/recommend.go:62]: That's ok for me.

Comment (yeya24) [reply pkg/test-infra/tiflash/recommend.go:62]: Yes, but just leave it here 😄

**Pull request merged**